# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name:

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [11]:
import requests

# Genius API endpoint for searching a song
base_url = "https://api.genius.com"

# Replace with your correct Client Access Token
access_token = "5kpybkI-3g9yM4gaTjieY0EjQhysKWro06gzvlKXvnFjT_M9NSWIbIRUWcThxjbM"

# Headers to include the token for authentication
headers = {
    'Authorization': f'Bearer {access_token}'
}

# Define the song you want to search for
song = "Birdhouse in Your Soul"
artist = "They Might Be Giants"

# Send a request to search for the song
search_url = f"{base_url}/search?q={song} {artist}"
response = requests.get(search_url, headers=headers)

# Check if the response is successful
if response.status_code == 200:
    data = response.json()
    song_info = data['response']['hits'][0]['result']
    song_url = song_info['url']
    print(f"Song found: {song_info['title']} by {song_info['primary_artist']['name']}")
    print(f"Song URL: {song_url}")
else:
    print("Failed to retrieve song data.")

Song found: Birdhouse in Your Soul by They Might Be Giants
Song URL: https://genius.com/They-might-be-giants-birdhouse-in-your-soul-lyrics


2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [15]:
import json
import os
from textblob import TextBlob

# Step 1: Fetch song data (if not already available)
# You should already have this data (replace with actual data if you already fetched it)
song_data = {
    'song_title': 'Birdhouse in Your Soul',
    'artist': 'They Might Be Giants',
    'lyrics': '''I'm your only friend
                 I'm not your only friend
                 But I'm a little glowing friend
                 But really I'm not actually your friend
                 But I am...'''
}

# Step 2: Save the song data to a JSON file
file_path = 'song_data.json'  # Adjust the file path if needed

# Check if the file already exists
if not os.path.exists(file_path):
    # Save the song data to a JSON file if it doesn't exist
    with open(file_path, 'w') as file:
        json.dump(song_data, file, indent=4)
    print(f"Song data has been saved to '{file_path}'")
else:
    print(f"File '{file_path}' already exists!")

# Step 3: Read the contents of the JSON file
# Load the song data from the JSON file
with open(file_path, 'r') as file:
    song_data = json.load(file)

# Step 4: Print only the song lyrics (not the entire dictionary)
print("\nSong Lyrics:\n", song_data['lyrics'])

# Step 5: Perform sentiment analysis using TextBlob
lyrics = song_data['lyrics']
blob = TextBlob(lyrics)
polarity = blob.sentiment.polarity  # Get the polarity score

# Step 6: Print the polarity score
print("\nPolarity Score:", polarity)

# Step 7: Analyze the sentiment based on polarity
if polarity > 0:
    sentiment = "positive"
elif polarity < 0:
    sentiment = "negative"
else:
    sentiment = "neutral"

# Step 8: Print the sentiment analysis result
print(f"\nThe sentiment of the lyrics is {sentiment}.")

# Optional: Add a comment based on the sentiment analysis
if polarity > 0:
    print("\nThis song's lyrics have a positive connotation.")
elif polarity < 0:
    print("\nThis song's lyrics have a negative connotation.")
else:
    print("\nThis song's lyrics have a neutral connotation.")

Song data has been saved to 'song_data.json'

Song Lyrics:
 I'm your only friend
                 I'm not your only friend
                 But I'm a little glowing friend
                 But really I'm not actually your friend
                 But I am...

Polarity Score: -0.046875

The sentiment of the lyrics is negative.

This song's lyrics have a negative connotation.


# Sentiment Analysis on Song Lyrics

## Overview

In this project, we performed **sentiment analysis** on the lyrics of the song **"Birdhouse in Your Soul"** by **They Might Be Giants** using the **TextBlob** library. The goal was to determine the emotional tone of the song, based on the polarity of its lyrics.

## Steps

1. **Retrieve Song Lyrics**: We fetched the song lyrics using the Genius API and stored them in a JSON file (`song_data.json`).
2. **Sentiment Analysis**: The song lyrics were then analyzed using **TextBlob**, which calculates the **polarity score** of the text.
3. **Interpret Results**: Based on the polarity score, we interpreted the sentiment of the song.

## Results

- **Song Found**: "Birdhouse in Your Soul" by They Might Be Giants
- **Polarity Score**: `-0.046875`
- **Sentiment**: Negative
- **Conclusion**: The lyrics have a slightly **negative** connotation, as indicated by the polarity score, which ranges from **-1.0** (very negative) to **1.0** (very positive). The small negative score suggests that the lyrics convey a mild melancholic or introspective mood.

## Analysis

- **Sentiment Interpretation**: Given that the polarity score is slightly negative but not extreme, it can be concluded that the lyrics have a **negative** connotation, but it is not overwhelmingly so. The song may convey themes of introspection or bittersweet emotions.

## Conclusion

The sentiment analysis has successfully provided insights into the emotional tone of the song's lyrics. With a **polarity score of -0.046875**, the lyrics lean slightly negative, which reflects a nuanced emotional message in the song.

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [ ]:
import requests
import json
from datetime import datetime
import time

# Replace with your own Genius API Token
GENIUS_API_TOKEN = 'yqBY32YOwJAVRQiWCo5wz9Wh78E86DfEI7d92t0yMcJitI9QiTzf0r-NBOph6rnH'

# Define the function to get lyrics and write them to a file
def write_song_from_genius_to_json(artist, song, filename, retries=3, delay=5):
    base_url = "https://api.genius.com"
    
    # Search for the song
    search_url = f"{base_url}/search"
    params = {'q': f"{song} {artist}"}
    
    for attempt in range(retries):
        print(f"Searching for {song} by {artist} (Attempt {attempt + 1}/{retries})")
        
        try:
            search_response = requests.get(search_url, params=params, headers={'Authorization': f'Bearer {GENIUS_API_TOKEN}'})
            
            if search_response.status_code == 200:
                search_data = search_response.json()
                
                # Check if any hit was found
                hits = search_data.get('response', {}).get('hits', [])
                if hits:
                    song_path = hits[0]['result']['path']
                    song_url = f"{base_url}{song_path}"

                    # Fetch lyrics from the song's page
                    lyrics_page_response = requests.get(song_url)
                    lyrics = extract_lyrics_from_page(lyrics_page_response.text)

                    if lyrics:
                        # Write lyrics to a file in JSON format
                        with open(filename, 'w', encoding='utf-8') as file:
                            json.dump({"artist": artist, "song": song, "lyrics": lyrics}, file, indent=4)
                        print(f"Lyrics for {song} by {artist} saved in {filename}")
                        return  # Exit the function once successful
                    else:
                        print(f"Lyrics not found for {song} by {artist}.")
                        break
                else:
                    print(f"No results found for {song} by {artist}.")

            else:
                print(f"Error: Could not search for {song} by {artist}. Status Code: {search_response.status_code}")
                print(f"Response: {search_response.text}")

        except requests.exceptions.RequestException as e:
            print(f"Request failed: {e}")

        # If we haven't succeeded after the current attempt, wait before retrying
        print(f"Retrying in {delay} seconds...\n")
        time.sleep(delay)

    # After all attempts, log the failure to a separate file for easier debugging
    log_failure(artist, song, retries)

# Function to extract lyrics from the song's page HTML
def extract_lyrics_from_page(page_html):
    start_index = page_html.find('<div class="lyrics">')
    end_index = page_html.find('</div>', start_index)

    if start_index != -1 and end_index != -1:
        lyrics_html = page_html[start_index:end_index]
        lyrics = ''.join(lyrics_html.split('<br/>'))  # Clean up the lyrics from HTML tags
        return lyrics.strip()
    return None

# Function to log the failure for later debugging
def log_failure(artist, song, retries):
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    with open("failed_requests_log.txt", 'a', encoding='utf-8') as log_file:
        log_file.write(f"{timestamp} - Failed to fetch lyrics for '{song}' by {artist} after {retries} attempts.\n")
    print(f"Failure logged for {song} by {artist}. Check 'failed_requests_log.txt' for details.\n")

# Test the function with 4 songs
songs = [
    ("Adele", "Someone Like You", "someone_like_you_genius.json"),
    ("Ed Sheeran", "Shape of You", "shape_of_you_genius.json"),
    ("Coldplay", "Yellow", "yellow_genius.json"),
    ("Taylor Swift", "Love Story", "love_story_genius.json")
]

for artist, song, filename in songs:
    write_song_from_genius_to_json(artist, song, filename)

Requesting: https://api.lyrics.ovh/v1/Adele/Someone Like You (Attempt 1/3)
Error: Could not fetch lyrics for Someone Like You by Adele. Status Code: 504
Response: <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx/1.18.0</center>
</body>
</html>

Retrying in 5 seconds...

Requesting: https://api.lyrics.ovh/v1/Adele/Someone Like You (Attempt 2/3)
Error: Could not fetch lyrics for Someone Like You by Adele. Status Code: 504
Response: <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx/1.18.0</center>
</body>
</html>

Retrying in 5 seconds...

Requesting: https://api.lyrics.ovh/v1/Adele/Someone Like You (Attempt 3/3)
Error: Could not fetch lyrics for Someone Like You by Adele. Status Code: 504
Response: <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
<hr><center>nginx/1.18.0</c

4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.